### PR TITLE
Fix incorrect optimization around lifted  0 - x  resulting in just  x

### LINKED ISSUE #7183: Incorrect code gen in VB when subtracting a nullable from another
From an internal bug from the support team:

__Given this:__
``` VB
Module Module1
  Sub Main()
    Dim a As Integer? = 1
    Dim b As Integer? = 0 - a
    Console.WriteLine(String.Format("a: {0}, b: {1}", a, b))
 End Sub
End Module
```

__Expected runtime result:__
```
a: 1, b: -1
```
__Actual runtime result:__
```
a: 1, b: 1
```

__Code gen:__
```
.method public static void  Main() cil managed
{
  .entrypoint
  .custom instance void [mscorlib]System.STAThreadAttribute::.ctor() = ( 01 00 00 00 ) 
  // Code size       71 (0x47)
  .maxstack  3
  .locals init ([0] valuetype [mscorlib]System.Nullable`1<int32> a,
           [1] valuetype [mscorlib]System.Nullable`1<int32> b,
           [2] valuetype [mscorlib]System.Nullable`1<int32> V_2)
  IL_0000:  nop
  IL_0001:  ldloca.s   a
  IL_0003:  ldc.i4.1
  IL_0004:  call       instance void valuetype [mscorlib]System.Nullable`1<int32>::.ctor(!0)
  IL_0009:  ldloca.s   a
  IL_000b:  call       instance bool valuetype [mscorlib]System.Nullable`1<int32>::get_HasValue()
  IL_0010:  brtrue.s   IL_001d
  IL_0012:  ldloca.s   V_2
  IL_0014:  initobj    valuetype [mscorlib]System.Nullable`1<int32>
  IL_001a:  ldloc.2
  IL_001b:  br.s       IL_0029
  IL_001d:  ldloca.s   a
  IL_001f:  call       instance !0 valuetype [mscorlib]System.Nullable`1<int32>::GetValueOrDefault()
  IL_0024:  newobj     instance void valuetype [mscorlib]System.Nullable`1<int32>::.ctor(!0)
  IL_0029:  stloc.1
  IL_002a:  ldstr      "a: {0}, b: {1}"
  IL_002f:  ldloc.0
  IL_0030:  box        valuetype [mscorlib]System.Nullable`1<int32>
  IL_0035:  ldloc.1
  IL_0036:  box        valuetype [mscorlib]System.Nullable`1<int32>
  IL_003b:  call       string [mscorlib]System.String::Format(string,
                                                              object,
                                                              object)
  IL_0040:  call       void [mscorlib]System.Console::WriteLine(string)
  IL_0045:  nop
  IL_0046:  ret
} // end of method Module1::Main
```

Notice that we don't actually subtract 'a' from anything.

### PR DESCRIPTION
**Pull request by VSadov**
That is a valid optimization for many binary operators like  + &  |  , but not for - .

Fixes #7183

### CODE REVIEW

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - please review

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_NullableHelpers.vb [hunk 1]
```diff
@@ -354,8 +354,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End If
 
             Select Case binaryOpKind
+                Case BinaryOperatorKind.Subtract
+                    If right.IsDefaultValueConstant Then
+                        Return left
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_NullableHelpers.vb]: Are we testing the code gen for this optimization?

**Comment by cston** [comment]
LGTM

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_NullableHelpers.vb (hunk 1 above)
Comment (VSadov) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_NullableHelpers.vb]: added a test for this.

**Comment by AlekseyTs** [comment]
Should we test unary minus on nullable int? Does VB do the right thing?

**Comment by VSadov** [comment]
@AlekseyTs  simple negation does no fall into this case as this is optimization for binary operators when one of the operands is 0 and another is not. 
C# is not affected here. It does not have this optimization at all. Perhaps makes sense to add it, but that is a different matter.

**Comment by AlekseyTs** [comment]
I am not saying that negation goes through this code path, just saying that it makes sense to verify that it doesn't have similar problem (for whatever reason). It also makes sense to verify VB's behavior regardless what optimizations it performs or doesn't perform.

**Comment by VSadov** [comment]
@AlekseyTs this fix _is_ for VB. I did check C# and the scenario worked fine. I also looked whether it performs this optimization correctly or does not do it at all. Turns out it does not do it at all.
-x is irrelevant because the root cause here is asymmetry of arguments in binary - . If the second arg is 0, then expression can be reduced to the first arg, but not the other way, unlike +, & and other binary expressions which are symmetrical.

**Comment by jaredpar** [comment]
I think we should add the corresponding C# tests as well.  They should be fairly simple to add and it guards us against making the same mistake in C# should we ever choose to add this optimization in the future.

**Comment by VSadov** [comment]
@jaredpar  added tests and actual optimizations to C#

**Comment by AlekseyTs** [comment]
To clarify, my concern was about checked(-x), which, if I remember correctly, is emitted as "0 - x". I don't remember whether local rewriter transforms it into a binary expression or emitter simply handles the transformation. Sorry for the confusion with the language, C# vs. VB.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by VSadov** [comment]
@AlekseyTs  checked -x is emitted as 0-x, because there is no such IL instruction (that would be neg.ovf ?), so it is emitted as checked(0-x) instead. This disconnect does have some interesting effects on optimizations like local slot scheduling, but lowering is not affected since lowering happens much earlier.
Interestingly, we could optimize 0-x as  -x, but that obviously can only be profitable when not in a checked context. Need to think about that.

**Pull request merged**